### PR TITLE
Allow retire_workers to run concurrently

### PR DESCRIPTION
**Pull request by fjetter**
Closes https://github.com/dask/distributed/issues/8053

The added test is rather superficial but it is simple and conveys what we want to work. There is also a little heavier test already available that does a similar thing

https://github.com/dask/distributed/blob/9a9d2c362971f7f31d0ba61ed6b4b151c6697b71/distributed/tests/test_active_memory_manager.py#L1230-L1269

running those locally didn't reveal any problems. if CI is not sadder than usual, I'm fine with this

Regarding the conditions: This will prohibit running replicate/rebalance while retire_workers is running and will prohibit running anything else while replicate/rebalance is running.
The alternative is to nuke these APIs entirely but the way I see it this complexity is not too bad. I considered factoring this out but this abstraction would look aweful since the semantics we need here are hard to generalize. This just felt unnecessary. Happy to discuss either option.

### CODE REVIEW

**Comment by fjetter** [comment]
FWIW I believe we also have pretty good consensus to nuke replicate/rebalance. Happy either way (from my POV this was actually _less_ work but I'm happy if somebody comes around deleting code)

**Comment by crusaderky** [comment]
I like this, but I think the synchronisation primitive is complex and hard to read enough to warrant a bit of encapsulation. I'm refactoring it now.

**Comment by crusaderky** [comment]
ready for review (assuming green tests)

File: distributed/asyncio.py
Comment (fjetter) [inline distributed/asyncio.py:1]: I suggest to make this module private. I would definitely recommend to use a name that is different from a stdlib module name

File: distributed/asyncio.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+    **Tip**
+
+    You can mix reentrant and non-reentrant owners; all you need to do is create an
+    owner that doesn't compare as equal to other instances of itself::
+
+        lock = RLock()
+
+        async def non_reentrant():
+            async with lock(object()):
```
Comment (fjetter) [inline distributed/asyncio.py:30]: the `object()` sentinel feels a bit odd. Why not just call it without an argument for this purpose?

File: distributed/tests/test_scheduler.py [hunk 2]
```diff
@@ -737,6 +737,31 @@ async def test_retire_workers_empty(s):
...
+    await asyncio.gather(
+        *(
+            s.replicate(keys=[fut.key for fut in futs], n=2),
+            s.rebalance(),
+            s.retire_workers(workers=[workers[0].address]),
+            s.retire_workers(workers=[workers[1].address]),
+        )
+    )
+    await c.gather(futs)
+    assert s.transition_counter == before
```
Comment (fjetter) [inline distributed/tests/test_scheduler.py:762]: This test doesn't care whether the three run concurrently or sequentially, does it? In fact, they could be replaced with no-ops and this test would pass.

I would like to see a test that actually verifies that the two retire_workers are indeed running concurrently.

File: distributed/asyncio.py (hunk 1 above)
Comment (fjetter) [reply distributed/asyncio.py:30]: I see how you are using it below. if the object is within a tuple, that makes more sense for this use case.

If we make the module private, I don't care about this

**Comment by crusaderky** [comment]
@fjetter all review comments have been addressed; ready for review again

**Comment by fjetter** [comment]
LGTM (I cannot _approve_ since I'm the original author)

**Pull request merged**